packages/api-fetch/src/test/index.js

```diff
@@ -113,83 +113,86 @@ describe( 'apiFetch', () => {
+			code: 'bad_request',
+			message: 'Bad Request',
 		} );
 	} );
 
-	it( 'should return invalid JSON error if no json response', () => {
-		window.fetch.mockReturnValue(
-			Promise.resolve( {
-				status: 200,
-			} )
-		);
+	it( 'should return invalid JSON error if no json response', async () => {
+		globalThis.fetch.mockResolvedValue( {
+			ok: true,
+			status: 200,
+			async json() {
+				return JSON.parse( '' );
+			},
+		} );
 
-		return apiFetch( { path: '/random' } ).catch( ( body ) => {
-			// eslint-disable-next-line jest/no-conditional-expect
-			expect( body ).toEqual( {
-				code: 'invalid_json',
-				message: 'The response is not a valid JSON response.',
-			} );
+		await expect( apiFetch( { path: '/random' } ) ).rejects.toEqual( {
+			code: 'invalid_json',
+			message: 'The response is not a valid JSON response.',
 		} );
 	} );
 
-	it( 'should return invalid JSON error if response is not valid', () => {
-		window.fetch.mockReturnValue(
-			Promise.resolve( {
-				status: 200,
-				json() {
-					return Promise.reject();
-				},
-			} )
-		);
+	it( 'should return invalid JSON error if response is not valid', async () => {
+		globalThis.fetch.mockResolvedValue( {
+			ok: true,
+			status: 200,
+			async json() {
+				return JSON.parse( '' );
+			},
+		} );
 
-		return apiFetch( { path: '/random' } ).catch( ( body ) => {
-			// eslint-disable-next-line jest/no-conditional-expect
-			expect( body ).toEqual( {
-				code: 'invalid_json',
-				message: 'The response is not a valid JSON response.',
-			} );
+		await expect( apiFetch( { path: '/random' } ) ).rejects.toEqual( {
+			code: 'invalid_json',
+			message: 'The response is not a valid JSON response.',
 		} );
 	} );
 
-	it( 'should return offline error when fetch errors', () => {
-		window.fetch.mockReturnValue( Promise.reject() );
+	it( 'should return offline error when fetch errors', async () => {
+		globalThis.fetch.mockRejectedValue(
+			new TypeError( 'Failed to fetch' )
+		);
 
-		return apiFetch( { path: '/random' } ).catch( ( body ) => {
-			// eslint-disable-next-line jest/no-conditional-expect
-			expect( body ).toEqual( {
-				code: 'fetch_error',
-				message: 'You are probably offline.',
-			} );
+		await expect( apiFetch( { path: '/random' } ) ).rejects.toEqual( {
+			code: 'fetch_error',
+			message: 'You are probably offline.',
 		} );
 	} );
 
 	it( 'should throw AbortError when fetch aborts', async () => {
-		const abortError = new Error();
-		abortError.name = 'AbortError';
-		abortError.code = 20;
-
-		window.fetch.mockReturnValue( Promise.reject( abortError ) );
+		globalThis.fetch.mockImplementation(
+			( path, { signal } ) =>
+				new Promise( ( _, reject ) => {
+					if ( ! signal ) {
+						reject( new Error( 'Expected signal as argument' ) );
+						return;
+					}
+
+					signal.throwIfAborted();
+					signal.addEventListener(
+						'abort',
+						( e ) => {
+							reject( e.target.reason );
+						},
+						{ once: true }
+					);
+				} )
+		);
```

Comment:
This is a realistic implementation of a `fetch` handler that:
1. Returns a promise that never resolves.
2. Aborts when someone calls `controller.abort()`.

The original version of the test just mocked `fetch` to reject with `AbortError` and then tested that this error was really thrown. The `signal` and the `controller` didn't play any role at all.